### ISSUE TITLE
From / Into for DynamicImage

### ISSUE THREAD

**Issue by NeoLegends** [image-rs/image#620]
I noticed there are conversion methods for `DynamicImage` into the various other image types, but none of these is consuming. For my tool I work with RGBA data only, so I convert every image I load into RGBA before usage. However, `DynamicImage::to_rgba()` always creates a copy, which would be unnecessary in the case I already am dealing with the RGBA subtype. I suggest that `From` / `Into` implementations for `DynamicImage`  be added that take care of the consuming conversion in a more optimized way.

I could file a PR if you like the change.

**Comment by nwin** [comment]
That is actually a little bit tricky to implement.

Furthermore I doubt that the gain will be huge, so if your program is slow that probably has another cause than one excess allocation per iteration.

**Comment by NeoLegends** [comment]
Why would it be? It might be a little verbose because one would have to do the matching, but not tricky.

Well, I consider this to be quite the waste of memory, to be honest. A 1920x1080 at 8bit Rgba already takes up 8 MB of memory. Just converting it into it's inner form will allocate another 8 megs that could've been avoided in some / most cases. And 1920x1080 is not a large image, if we're talking about processing photographs, we're talking about 15-25MBs per image each.

**Comment by nwin** [comment]
I would be tricky, because you would need to operate on raw bytes, drop all abstraction and handle the subpixels manually.

I wouldn’t be so sure about the waste of memory either, you assume that you can just resize a vector without reallocating and moving which isn’t always true. The only difference would be that you drop you additional image faster. Have a look: https://is.gd/sA2jMm

It would probably be more efficient if you would enable the api to reuse a buffer.

**Comment by NeoLegends** [comment]
Oh, seems like we're talking about different things here! 😆 

I just want a convenient and more efficient way to convert `DynamicImage` (the enum) into it's inner variants (`RgbImage`, `RgbaImage`, etc.). The actual conversion in the case where the source type and the target type of the conversion does not match up, can remain.

Basically I want something like [this](https://docs.rs/image/0.12.2/image/enum.DynamicImage.html#method.to_rgba) in a consuming fashion that is optimized for the case when we're already dealing with the `DynamicImage::ImageRgba8` case. Like this pseudo-code:

```
impl From<DynamicImage> for RgbaImage {
    fn from(val: DynamicImage) -> Self {
        if let DynamicImage::ImageRgba8(img) = val {
            val
        } else {
            val.to_rgba8()
        }
    }
}
```

This shouldn't need any subpixel-handling or raw-bytes operations that aren't already there.

**Comment by HeroicKatora** [comment]
This was implemented in #1096 as dedicated methods but I missed closing this.
